autosklearn/automl.py

```diff
@@ -619,11 +656,22 @@ def fit(
 
         # Wait until the ensemble process is finished to avoid shutting down
         # while the ensemble builder tries to access the data
-        if self._proc_ensemble is not None and self._ensemble_size > 0:
-            self._proc_ensemble.join()
-            self.ensemble_performance_history = self._proc_ensemble.get_ensemble_history()
-
-        self._proc_ensemble = None
+        if proc_ensemble is not None:
+            # Here we signal the ensemble process to finish
+            # That is, no more iterations are allowed
+            self.ensemble_event_killer.set()
+
+            # We wait for self.ensemble_timeout_patience seconds for a job to finish
+            # Notice we provided a kill signal and also the maximum allowed time, so
+            # it is likely that the TimeoutError won't happen. In case it does, we notify
+            # the user
+            try:
+                dask.distributed.wait(proc_ensemble, timeout=self.ensemble_timeout_patience)
+                self.ensemble_performance_history = proc_ensemble.result()
+            except dask.distributed.TimeoutError:
+                self._logger.error("The ensemble process unexpectedly hanged..."
+                                   "Please check the log file for errors.")
+                self._dask_client.cancel(proc_ensemble)
```

Comment:
Does this guarantee immediate termination?

Reply:
It cancels the job, but actually the immediate termination is guaranteed when the client is closed.

This is to clean up resource-usage-wise. I don;t expect to reach this scenario as we send a kill event AND have a internal time tracking. But this is a safety measure to make sure that in the worst case scenario we clean up the submitted jobs.

BTW, it is very bad IF we get to this point, because the ensemble history will be empty. So only on a real bug we should reach here, and the message will be a good indicator for this.